subsys/bluetooth/host/scan.c

```diff
@@ -24,14 +25,25 @@
 #include "id.h"
 
 #include "common/bt_str.h"
+#include "scan.h"
+#include <stdint.h>
+#include <string.h>
 
 #define LOG_LEVEL CONFIG_BT_HCI_CORE_LOG_LEVEL
 #include <zephyr/logging/log.h>
 LOG_MODULE_REGISTER(bt_scan);
 
+struct scanner_state {
+	ATOMIC_DEFINE(scan_flags, SCAN_ENABLED_REASON_NUM_FLAGS);
+	struct bt_le_scan_param explicit_scan_param;
+	struct bt_le_scan_param used_scan_param;
+};
+
 static bt_le_scan_cb_t *scan_dev_found_cb;
 static sys_slist_t scan_cbs = SYS_SLIST_STATIC_INIT(&scan_cbs);
 
+static struct scanner_state m_scanner_state;
```

Comment:
It kind of makes sense to have a scanner state in `scan.c`, but we also keep some scan state in the `bt_dev` now, so we basically have 2 scan states now?

Reply:
As I see it, bt_dev keeps track of which roles that are running. I think it makes sense to keep `BT_DEV_SCANNING` there. However, I can move that to the scanner state, as well.
The logic related to `BT_DEV_SCAN_LIMITED`, I would like to address in a follow-up.

I liked the idea of having more self-contained role-modules. Having the state inside the module prevents other modules on relying on the internals of the scanner. This in turn should make the code more maintainable. But I can add the state to the bt_dev instead, if that is more consistent with other roles.